src/pybind/mgr/cephadm/migrations.py

```diff
@@ -0,0 +1,116 @@
+
+
+
+class Migrations:
+    def __init__(self, mgr: "CephadmOrchestrator"):
+        self.mgr = mgr
+
+        self.initialize()
+
+    def initialize(self):
+        if self.mgr.migration_current is None:
+            self.set(0)
+        self.migrate()
+
+    def set(self, val):
+        self.mgr.set_module_option('migration_current', val)
+        self.mgr.migration_current = val
+
+    def is_migration_ongoing(self):
+        return self.mgr.migration_current != LAST_MIGRATION
+
+    def verify_no_migration(self):
+        if self.is_migration_ongoing():
+            raise OrchestratorError("cephadm migration still ongoing. Please wait, until the migration is complete.")
+
+    def migrate(self):
+        if self.mgr.migration_current == 0:
+            if self.migrate_0_1():
+                self.set(1)
+
+    def migrate_0_1(self):
+        """
+        Migration 0 -> 1
+        New scheduler that takes PlacementSpec as the bound and not as recommendation.
+
+        Which means, we have to make sure, we're not removing any daemons directly after
+        upgrading to the new scheduler.
+
+        There is a potential race here:
+        1. user updates his spec to remove daemons
+        2. mgr gets upgrades to new scheduler, before the old scheduler removed the daemon
+        3. now, we're converting the spec to explicit placement, thus reverting (1.)
+        I think this is ok.
+        """
+
+        def interesting_specs() -> Iterator[ServiceSpec]:
+            for s in self.mgr.spec_store.specs.values():
+                if s.unmanaged:
+                    continue
+                p = s.placement
+                if p is None:
+                    continue
+                if p.count is None:
+                    continue
+                if not p.hosts and not p.host_pattern and not p.label:
+                    continue
+                yield s
+
+        def convert_to_explicit(spec: ServiceSpec):
+            placements = HostAssignment(
+                spec=spec,
+                get_hosts_func=self.mgr._get_hosts,
+                get_daemons_func=self.mgr.cache.get_daemons_by_service
+            ).place()
+
+            existing_daemons = self.mgr.cache.get_daemons_by_service(spec.service_name())
+
+            # We have to migrate, only if the new scheduler would remove daemons
+            if len(placements) >= len(existing_daemons):
+                return
+
+            old_hosts = {h.hostname: h for h in spec.placement.hosts}
+            new_hosts = [
+                old_hosts[d.hostname] if d.hostname in old_hosts else HostPlacementSpec(hostname=d.hostname, network='', name='')
+                for d in existing_daemons
+            ]
+
+            new_placement = PlacementSpec(
+                hosts=new_hosts,
+                count=spec.placement.count
+            )
+
+            new_spec = ServiceSpec.from_json(spec.to_json())
+            new_spec.placement = new_placement
```

Comment:
I can also see an issue where a user would apply a cluster.yaml only when changing a single service. Since we converted the placement_glob(or pattern) to an explicit placement spec, they probably have to export and overwrite their existing cluster.yaml. 

I think this should be recommended somehow.

Reply:
Right, that's a big problem but I didn't have a better answer.

Reply:
Having previews for all services would be a first step I think.. If placements would be changed _completely_ we could atleast raise a warning before applying it.

Reply:
both schedulers behave very similar when it comes to assigning hosts. It's just the old assigned more hosts than the new one.

Reply:
I'm still struggling with this concept, but I suppose it's similar to k8s `nodeName` ...
https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodename

Reply:
yes. but what's the alternative? what if the user wrongly created a placement like 

```yaml
service_type: mon
placement:
    label: mymons
    count=5
```
but only a single host was labled as `mymons`? should we simply remove all the mons? I'm also against being more elaborate here as this increases the likelihood of bugs. I'd be really in favor of keeping this as simple as possible to stay safe.